webtransport/README.md

```diff
@@ -0,0 +1,60 @@
+
+Authors: [@marten-seemann]
+
+Interest Group: TODO
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level
+and spec status.
+
+## Introduction
+
+WebTransport is a way for browsers to establish a stream-multiplexed connection to servers that allows bidirectional streaming.
+
+The WebTransport protocol is currently under development at the IETF. Chrome has implemented and shipped support for [draft-02](https://datatracker.ietf.org/doc/draft-ietf-webtrans-http3/).
+
+The most exciting feature for libp2p (other than the improved performance and hole punching success rates that QUIC gives us) is that the W3C added a browser API allowing browsers to establish connections to nodes with self-signed certificates, provided they know the hash of the certificate in advance: `[serverCertificateHashes](https://www.w3.org/TR/webtransport/#dom-webtransportoptions-servercertificatehashes)`. This API is already [implemented in Chrome](https://chromestatus.com/feature/5690646332440576). Firefox is working on a WebTransport implementation and is [likely to implement](https://github.com/mozilla/standards-positions/issues/167#issuecomment-1015951396) `serverCertificateHashes` as well.
+
+## Certificates
+
+Server nodes can choose between 2 modes of operation:
+1. They possess a certificate that's accepted by the WebPKI (i.e. a certificate signed by a well-known certificate authority). Due to the p2p nature of libp2p, obtaining such a certificate will only be practical for a small number of nodes.
+2. They use a self-signed certificate. According to the [w3c WebTransport certification](https://www.w3.org/TR/webtransport/), the validity of the certificate MUST be at most 14 days. Nodes then include the hash of one (or more) certificates in their multiaddr (see [#addressing]).
+
+When using self-signed certificates, nodes need to take care to regularly renew their certificate. In the following, the RECOMMENDED logic for rolling certificates is described. At first boot of the node, it creates a self-signed certificate with a validity of 14 days, and publishes a multiaddr containing the hash of that certificate. After 10 days, the node prepares the next certificate, setting the `NotBefore` date of that certificate to the expiration date (or shortly before that) of the first certificate, and an expiration of 14 days after that. The node continues using the old certificate until its expiry date, but it already advertises a multiaddr containing both certificate hashes. This way, clients will be able to connect to the node, even if they cache the multiaddr for multiple days.
+
+## Addressing
+
+Webtransport multiaddresses are composed of a QUIC multiaddr, followed by `/webtransport` and a list of multihashes of the certificates that the server uses.
+Examples:
+* `/ip4/1.2.3.4/udp/443/quic/webtransport/`
+* `/ip6/fe80::1ff:fe23:4567:890a/udp/1234/quic/webtransport/<hash1><hash2><hash3>`
```

Comment:
might be useful to describe what the string serialization of hashes here are. would it be the base58 direct encoding of the multihash, or the CID encoding with a multibase prefix?

Reply:
Probably multibase, then we don't have to specify the encoding.

Looks like the encoding I suggested above doesn't work though. We can't have an optional hash argument.
I'm also not sure how to include the path (as you suggest in https://github.com/libp2p/specs/pull/404/files#r842506567) in the address, as `/webtransport` can't have 2 values: `/ip4/1.2.3.4/udp/443/quic/webtransport/<hash>/<encoded url>`.

Maybe the best solution is to introduce 2 new multiaddr components:
`/ip4/1.2.3.4/udp/443/quic/webtransport/certhashes/<hash1><hash2>/url/<encoded url>`, or even `/ip4/1.2.3.4/udp/443/quic/webtransport/certhash/<hash1>/certhash/<hash2>/url/<encoded url>`. wdyt?

Reply:
* I don't have a strong feeling about concatenation of hashes versus having each one in a separate `certhash` subcomponent.  Concatenation is probably less work for existing parsers that don't always think about repeated components.
* having a separate `/url/<encoded url>` seems reasonable as an optional over-ride to well-known.

Reply:
@mriise We definitely need to support 2 certs, but I don't see a reason to impose any restrictions that go further than the general length limit we have for multiaddrs.

The reason is the following: The way that `serverCertificateHashes` works is NOT by the browser telling the server "I'd like to see a cert with the following hash". Instead, the browser has a (list of) certificate hashes that it is willing to accept.

The node would generate one certificate for the first 2 weeks, another certificate for the 2 weeks after that, and so on. If you expect that peers will save your address for a prolonged period of time, you could advertise more than 2 hashes, but I expect that too be very rare.

Reply:
> We definitely need to support 2 certs, but I don't see a reason to impose any restrictions that go further than the general length limit we have for multiaddrs.

This makes sense as a default definitely. Should there be checks to make sure they aren't the same hash to prevent people from only excepting one anyways?  

I guess a logical maximum for certs would be 14, 1 per day.

Reply:
> We definitely need to support 2 certs

sorry for barging in, but I still don't understand why at least two certs are needed and not just one. could you explain or drop a link to spec or something (e.g. https://w3c.github.io/webtransport/#dom-webtransportoptions-servercertificatehashes)

Reply:
See line 28 of this document.

Reply:
I think for announcement only one cert is needed. The node needs to be capable of accepting both of them.
So few days before cert expiry, the node start announcing cert2, but it still accepts connections coming with cert1.

In essence during that period the node has two valid multiaddresses, which I think is fine.

Reply:
Ahh, the server can only support one of them, and client has to specify both. Feel free to ignore what I said.